### ISSUE TITLE
more precise semantic highlighting

### ISSUE THREAD

**Issue by Trass3r** [clangd/clangd#494]
With proper modifiers: https://github.com/Trass3r/llvm-project/commits/semanticmods

**Comment by kirillbobyrev** [comment]
Yeah this has been on our radars for the past couple of months but I didn't get time to implement this, thank you very much for bringing it up and also working on this!

I didn't look into the patch very close but it looks nice, would you like to submit it to [Phabicator](https://reviews.llvm.org/) for a review?

**Comment by Trass3r** [comment]
Yes though unfortunately it doesn't really make a big difference. Few themes actually make use of it, only some color const differently.

**Comment by HighCommander4** [comment]
> Yes though unfortunately it doesn't really make a big difference. Few themes actually make use of it, only some color const differently.

In vscode, you can specify your own colors using `editor.tokenColorCustomizationsExperimental` as shown [here](https://code.visualstudio.com/updates/v1_41#_semantic-tokens-provider).

**Comment by puremourning** [comment]
> Yes though unfortunately it doesn't really make a big difference. Few themes actually make use of it, only some color const differently.

What’s the difference to the client implementer? I’m working on YCM suppprt for semantic tokens.

**Comment by Trass3r** [comment]
Information about this is a bit hidden. This explains it pretty well, it was added in 1.44: https://github.com/microsoft/vscode/wiki/Semantic-Highlighting-Overview#as-a-theme-author-do-i-need-to-change-my-theme-to-make-it-work-with-semantic-highlighting

So you can style the code based on semantic tokens (which is not widely used yet) or as a fallback based on textmate scopes (which may require a custom [mapping](https://code.visualstudio.com/api/language-extensions/semantic-highlight-guide#semantic-token-scope-map) in vscode-clangd package.json). But themes actually need to colorize things based on those modifiers and not just use the same color: https://github.com/material-theme/vsc-material-theme/commit/d322b22fdc1312077f9485f36af561035296c4c6

@kirillbobyrev Any ideas for simplifying the code?
Like e.g. ideally there should be a single decl.isStatic() check to determine that modifier.

**Comment by kirillbobyrev** [comment]
> @kirillbobyrev Any ideas for simplifying the code?
> Like e.g. ideally there should be a single decl.isStatic() check to determine that modifier.

I can't figure out something out-of-the box right now, I'd have to take a closer look (will be on vacation the next week though).

As for the usefulness, I think semantic highlighting will become more useful and more themes will support it soon-ish because right now it just stabilized, right? This is a complete speculation, though.

**Comment by Trass3r** [comment]
> I didn't look into the patch very close but it looks nice, would you like to submit it to [Phabicator](https://reviews.llvm.org/) for a review?

@kirillbobyrev Feel free to just add comments to https://github.com/Trass3r/llvm-project/commit/ff186df1ba85461743b9795bd280bbcf7c20b039 or in here while this is still WIP.

**Comment by Trass3r** [comment]
Looks like there's previous/related work: https://reviews.llvm.org/D77702

**Comment by HighCommander4** [comment]
Sam has written a nice set of patches to add various token modifiers and make the semantic highlightings more expressive:

 * https://reviews.llvm.org/D77811
 * https://reviews.llvm.org/D95701
 * https://reviews.llvm.org/D95706

These patches accomplish everything that https://reviews.llvm.org/D77702 did and more (thanks Sam!), and also handle many of the modifiers present in @Trass3r's patch in this issue (though there are some remaining ones in that patch, like `DefaultLibrary`, which it might make sense to spin out into a separate patch).

**Comment by HighCommander4** [comment]
> In vscode, you can specify your own colors using `editor.tokenColorCustomizationsExperimental` as shown [here](https://code.visualstudio.com/updates/v1_41#_semantic-tokens-provider).

In newer vscode versions, this is no longer experimental, and the setting is `editor.semanticTokenColorCustomizations` (described [here](https://code.visualstudio.com/updates/v1_45#_semantic-token-styling)).
